From e32ad5e03be7b64fd8b672e93d718a709861edd2 Mon Sep 17 00:00:00 2001
From: Chen Minqiang <ptpt52@gmail.com>
Date: Sat, 5 Jun 2021 14:16:16 +0800
Subject: [PATCH] net: phy: realtek: RTL8211FS support for fiber mode

This introduce "mode-fiber" property to indicate the phy working in
fiber mode.

&mdio {
	ephy7: ethernet-phy@7 {
		reg = <7>;
		mode-fiber;
	};
};

Signed-off-by: Chen Minqiang <ptpt52@gmail.com>
---
 drivers/net/phy/realtek.c | 123 +++++++++++++++++++++++++++++++++++++-
 1 file changed, 122 insertions(+), 1 deletion(-)

diff --git a/drivers/net/phy/realtek.c b/drivers/net/phy/realtek.c
index 29ff713..e7cf005 100644
--- a/drivers/net/phy/realtek.c
+++ b/drivers/net/phy/realtek.c
@@ -9,6 +9,8 @@
  */
 #include <linux/bitops.h>
 #include <linux/phy.h>
+#include <linux/sfp.h>
+#include <linux/phylink.h>
 #include <linux/module.h>
 #include <linux/delay.h>
 
@@ -62,6 +64,10 @@ MODULE_DESCRIPTION("Realtek PHY driver");
 MODULE_AUTHOR("Johnson Leung");
 MODULE_LICENSE("GPL");
 
+struct realtek_priv {
+	bool mode_fiber;
+};
+
 static int rtl821x_read_page(struct phy_device *phydev)
 {
 	return __phy_read(phydev, RTL821x_PAGE_SELECT);
@@ -547,6 +553,119 @@ static int rtlgen_resume(struct phy_device *phydev)
 	return ret;
 }
 
+static int rtl8211f_config_aneg(struct phy_device *phydev)
+{
+	struct realtek_priv *priv = phydev->priv;
+	int ret;
+
+	if (priv->mode_fiber) {
+		ret = genphy_c37_config_aneg(phydev);
+		if (ret < 0)
+			return ret;
+	} else {
+		ret = genphy_config_aneg(phydev);
+		if (ret < 0)
+			return ret;
+	}
+
+	/* Quirk was copied from vendor driver. Unfortunately it includes no
+	 * description of the magic numbers.
+	 */
+	if (phydev->speed == SPEED_100 && phydev->autoneg == AUTONEG_DISABLE) {
+		phy_write(phydev, 0x17, 0x2138);
+		phy_write(phydev, 0x0e, 0x0260);
+	} else {
+		phy_write(phydev, 0x17, 0x2108);
+		phy_write(phydev, 0x0e, 0x0000);
+	}
+
+	return 0;
+}
+
+static int rtl8211f_read_status(struct phy_device *phydev)
+{
+	struct realtek_priv *priv = phydev->priv;
+
+	if (priv->mode_fiber) {
+		/* Handle (Fiber) SGMII to RGMII mode */
+		return genphy_c37_read_status(phydev);
+	}
+
+	return rtlgen_read_status(phydev);
+}
+
+static int rtl8211f_sfp_insert(void *upstream, const struct sfp_eeprom_id *id)
+{
+	__ETHTOOL_DECLARE_LINK_MODE_MASK(rtl8211f_support) = { 0, };
+	__ETHTOOL_DECLARE_LINK_MODE_MASK(support) = { 0, };
+	struct phy_device *phydev = upstream;
+	phy_interface_t iface;
+
+	phylink_set(rtl8211f_support, 1000baseX_Full);
+	/* AT803x only support 1000baseX but SGMII works fine when module runs
+	 * at 1Gbit.
+	 */
+	phylink_set(rtl8211f_support, 1000baseT_Full);
+
+	sfp_parse_support(phydev->sfp_bus, id, support);
+
+	// Limit to interfaces that both sides support
+	linkmode_and(support, support, rtl8211f_support);
+
+	if (linkmode_empty(support))
+		goto unsupported_mode;
+
+	iface = sfp_select_interface(phydev->sfp_bus, support);
+
+	if (iface != PHY_INTERFACE_MODE_SGMII &&
+	    iface != PHY_INTERFACE_MODE_1000BASEX)
+		goto unsupported_mode;
+
+	dev_info(&phydev->mdio.dev, "SFP interface %s", phy_modes(iface));
+
+	return 0;
+
+unsupported_mode:
+	dev_info(&phydev->mdio.dev, "incompatible SFP module inserted;"
+		 "Only SGMII at 1Gbit/1000BASEX are supported!\n");
+	return -EINVAL;
+}
+
+static const struct sfp_upstream_ops rtl8211f_sfp_ops = {
+	.attach = phy_sfp_attach,
+	.detach = phy_sfp_detach,
+	.module_insert = rtl8211f_sfp_insert,
+};
+
+static int rtl8211f_probe(struct phy_device *phydev)
+{
+	struct device *dev = &phydev->mdio.dev;
+	struct device_node *np = dev->of_node;
+	bool mode_fiber = false;
+	struct realtek_priv *priv;
+	int ret;
+
+	mode_fiber = of_property_read_bool(np, "mode-fiber");
+
+	if (mode_fiber) {
+		ret = phy_sfp_probe(phydev, &rtl8211f_sfp_ops);
+		if (ret < 0)
+			return ret;
+	}
+
+	/* disable response to mdio phy0 commnad */
+	phy_modify_paged_changed(phydev, 0xa43, 24, BIT(13), 0);
+
+	priv = devm_kzalloc(dev, sizeof(*priv), GFP_KERNEL);
+	if (!priv)
+		return -ENOMEM;
+
+	priv->mode_fiber = mode_fiber;
+	phydev->priv = priv;
+
+	return 0;
+}
+
 static struct phy_driver realtek_drvs[] = {
 	{
 		PHY_ID_MATCH_EXACT(0x00008201),
@@ -620,10 +739,12 @@ static struct phy_driver realtek_drvs[] = {
 	}, {
 		PHY_ID_MATCH_EXACT(0x001cc916),
 		.name		= "RTL8211F Gigabit Ethernet",
+		.probe		= rtl8211f_probe,
 		.config_init	= &rtl8211f_config_init,
 		.ack_interrupt	= &rtl8211f_ack_interrupt,
-		.read_status	= rtlgen_read_status,
+		.read_status	= rtl8211f_read_status,
 		.config_intr	= &rtl8211f_config_intr,
+		.config_aneg	= rtl8211f_config_aneg,
 		.suspend	= genphy_suspend,
 		.resume		= genphy_resume,
 		.read_page	= rtl821x_read_page,
-- 
2.17.1

